tachyon/math/matrix/matrix_operations.h

```diff
@@ -0,0 +1,160 @@
+#ifndef TACHYON_MATH_MATRIX_MATRIX_OPERATIONS_H_
+#define TACHYON_MATH_MATRIX_MATRIX_OPERATIONS_H_
+
+#include "tachyon/base/openmp_util.h"
+#include "tachyon/math/matrix/matrix_types.h"
+
+namespace tachyon::math {
+
+// The motivation of the variants of |MulMatVec()| is as follows:
+
+// NOTE(chokobole): Eigen matrix multiplication has a computational
+// overhead unlike naive matrix multiplication.
+//
+// Example: Multiplying a 2x2 matrix by a 2x1 vector:
+//
+// +----+----+   +----+
+// | m₀ | m₁ | * | v₀ |
+// +----+----+   +----+
+// | m₂ | m₃ |   | v₁ |
+// +----+----+   +----+
+//
+// The operations involved in this multiplication are as follows:
+//
+// 1 * 1
+// 1 * 1
+// m₀ * v₀
+// m₀v₀ + 0
+// m₂ * v₀
+// m₂v₀ + 0
+// m₁ * v₁
+// m₁v₁ + m₀v₀
+// m₃ * v₁
+// m₃v₁ + m₂v₀
+// m₁v₁ + m₀v₀ * 1
+// m₁v₁ + m₀v₀ + 0
+// m₃v₁ + m₂v₀ * 1
+// m₃v₁ + m₂v₀ + 0
+
+template <typename Derived, typename Derived2,
+          typename F = typename Derived::Scalar>
+math::Vector<F> MulMatVec(const Eigen::MatrixBase<Derived>& matrix,
+                          const Eigen::MatrixBase<Derived2>& vector) {
+  static_assert(std::is_same_v<F, typename Derived2::Scalar>);
+
+  math::Vector<F> ret = math::Vector<F>::Constant(vector.size(), F::Zero());
+  OPENMP_PARALLEL_FOR(Eigen::Index i = 0; i < matrix.rows(); ++i) {
+    for (Eigen::Index j = 0; j < matrix.cols(); ++j) {
+      ret[i] += matrix(i, j) * vector[j];
+    }
+  }
+  return ret;
+}
+
+template <typename Derived, typename Derived2,
+          typename F = typename Derived::Scalar>
+math::Vector<F> MulMatVecSerial(const Eigen::MatrixBase<Derived>& matrix,
+                                const Eigen::MatrixBase<Derived2>& vector) {
+  static_assert(std::is_same_v<F, typename Derived2::Scalar>);
+
+  math::Vector<F> ret = math::Vector<F>::Constant(vector.size(), F::Zero());
+  for (Eigen::Index i = 0; i < matrix.rows(); ++i) {
+    for (Eigen::Index j = 0; j < matrix.cols(); ++j) {
+      ret[i] += matrix(i, j) * vector[j];
+    }
+  }
+  return ret;
+}
+
+template <typename Derived, typename Derived2, enum Eigen::AccessorLevels Level,
+          typename F = typename Derived::Scalar>
+math::Vector<F> MulMatVec(
+    const Eigen::MatrixBase<Derived>& matrix,
+    const Eigen::DenseCoeffsBase<Derived2, Level>& vector) {
+  static_assert(std::is_same_v<F, typename Derived2::Scalar>);
+
+  math::Vector<F> ret = math::Vector<F>::Constant(vector.size(), F::Zero());
```

Comment:
It would be good if it matches the max value of the for statement.

OPENMP_PARALLEL_FOR(Eigen::Index i = 0; i < `matrix.rows()`; ++i)

```suggestion
  math::Vector<F> ret = math::Vector<F>::Constant(matrix.rows(), F::Zero());
```